File: atcoder/trainings/abc328_e.py
class UnionFind:
    def __init__(self, size):
        self.rank = [0] * size
        self.parent = list(range(size))

    def unite(self, id_x, id_y):
        x_parent = self.find(id_x)
        y_parent = self.find(id_y)
        if x_parent == y_parent:
            return

        if self.rank[x_parent] > self.rank[y_parent]:
            self.parent[y_parent] = x_parent
        else:
            self.parent[x_parent] = y_parent
            if self.rank[x_parent] == self.rank[y_parent]:
                self.rank[y_parent] += 1

    def find(self, id_x):
        if self.parent[id_x] != id_x:
            self.parent[id_x] = self.find(self.parent[id_x])
        return self.parent[id_x]

    def same(self, id_x, id_y):
        return self.find(id_x) == self.find(id_y)

    def size(self, i):
        return self.rank[self.find(i)]


def dfs(i, arr, n, m, k, uvw, ans):
    if len(arr) == n - 1:
        uf = UnionFind(n)
        tmp = 0
        ok = True
        for j in arr:
            u, v, w = uvw[j]
            if uf.same(u, v):
                ok = False
                break
            uf.unite(u, v)
            tmp = (tmp + w) % k
        if ok:
            ans[0] = min(ans[0], tmp)
    else:
        if i == m:
            return
        dfs(i + 1, arr, n, m, k, uvw, ans)
        dfs(i + 1, arr + [i], n, m, k, uvw, ans)


n, m, k = map(int, input().split())
uvw = [list(map(int, input().split())) for _ in range(m)]
for i in range(m):
    uvw[i][0] -= 1
    uvw[i][1] -= 1

ans = [10**18]
dfs(0, [], n, m, k, uvw, ans)
print(ans[0])

File: atcoder/trainings/abc328_e.rb
# frozen_string_literal: true

# https://atcoder.jp/contests/abc328/tasks/abc328_e
#
# ■考えたこと
#
# ■解説
#
class UnionFind
  attr_accessor :rank, :parent

  def initialize(size)
    @rank = Array.new(size, 0)
    @parent = Array.new(size, &:itself)
  end

  def unite(id_x, id_y)
    x_parent = parent(id_x)
    y_parent = parent(id_y)
    return if x_parent == y_parent

    if @rank[x_parent] > @rank[y_parent]
      @parent[y_parent] = x_parent
    else
      @parent[x_parent] = y_parent
      @rank[y_parent] += 1 if @rank[x_parent] == @rank[y_parent]
    end
  end

  def parent(id_x)
    @parent[id_x] == id_x ? id_x : (@parent[id_x] = parent(@parent[id_x]))
  end

  def same?(id_x, id_y)
    parent(id_x) == parent(id_y)
  end

  def size(i)
    @rank[parent(i)]
  end
end

@n, @m, @k = gets.split.map(&:to_i)
@uvw = @m.times.map do
  u, v, w = gets.split.map(&:to_i)
  [u - 1, v - 1, w]
end

# m-1C7の全パターンを列挙して調べる
@ans = 10**18
def dfs(i, arr)
  if arr.size == @n - 1
    uf = UnionFind.new(@n)
    tmp = 0
    ok = true
    arr.each do |j|
      u, v, w = @uvw[j]
      # 連結になっている場合は全域木にならないので、そのパターンは除外
      ok = false if uf.same?(u, v)

      uf.unite(u, v)
      tmp = (tmp + w) % @k
    end
    @ans = [@ans, tmp].min if ok
  else
    # mは最大でm-1C7になるのでiがmになったら終了
    return if i == @m

    # m-1C7のパターンのうち、このuvwの組みを選ぶか選ばないか
    dfs(i + 1, arr)
    dfs(i + 1, Array.new(arr) << i)
  end
end

dfs(0, [])
puts @ans
